test/e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go

```diff
@@ -611,23 +612,26 @@ func waitForReplicas(ctx context.Context, deploymentName, namespace string, cs c
 
 func ensureDesiredReplicasInRange(ctx context.Context, deploymentName, namespace string, cs clientset.Interface, minDesiredReplicas, maxDesiredReplicas int, timeout time.Duration) {
 	interval := 60 * time.Second
+	errBelow := fmt.Errorf("number of replicas below target")
+	errAbove := fmt.Errorf("number of replicas above target")
 	err := wait.PollUntilContextTimeout(ctx, interval, timeout, true, func(ctx context.Context) (bool, error) {
 		deployment, err := cs.AppsV1().Deployments(namespace).Get(ctx, deploymentName, metav1.GetOptions{})
 		if err != nil {
-			framework.Failf("Failed to get replication controller %s: %v", deployment, err)
+			framework.Logf("Failed to get replication controller %s: %v", deployment, err)
+			return false, err
 		}
 		replicas := int(deployment.Status.ReadyReplicas)
 		framework.Logf("expecting there to be in [%d, %d] replicas (are: %d)", minDesiredReplicas, maxDesiredReplicas, replicas)
 		if replicas < minDesiredReplicas {
-			return false, fmt.Errorf("number of replicas below target")
+			return false, errBelow
 		} else if replicas > maxDesiredReplicas {
-			return false, fmt.Errorf("number of replicas above target")
+			return false, errAbove
 		} else {
 			return false, nil // Expected number of replicas found. Continue polling until timeout.
 		}
 	})
 	// The call above always returns an error, but if it is timeout, it's OK (condition satisfied all the time).
-	if wait.Interrupted(err) {
+	if err != errBelow && err != errAbove {
```

Comment:
why changing this? it is only sucesss if this condition matches

Reply:
I'm actually testing it in this pr, but hope you can help me dig it.

Before I just change the failf to return an error, and the test still fails.
See this comment: https://github.com/kubernetes/kubernetes/pull/122446#issuecomment-1867427179
This error is clearly also a timeout error. However, it seems that it's throw by client-go and will not match `Interrupted` here @aojea .

Reply:
It's reproducible https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/122446/pull-kubernetes-e2e-autoscaling-hpa-cm/1738166784302780416
I think when we got a timeout during `cs.AppsV1().Deployments(namespace).Get()`, the timeout returned by client-go doesn't match with `wait.Interrupted`.

Reply:
> any other error that is not timeout should fail

I think client-go returned error is a small probability event. And as this function only meant to ensure the number of replicas. It should be ok to ignore that. Also, I can't think of a better way to deal with this error(Wait(n=1) would exceed context deadline) right now.

Reply:
if an unexpected error is not expected we must fail, we must be strict on the tests, the end goal of the test is to ensure the code is reliable, is not to just pass, if we relax conditions this translates to unreliable systems

> client rate limiter Wait returned an error: context deadline exceeded

then if this error may be legit, we have to add that error to the condition, but not relax the criteria